rs/xnet/payload_builder/src/certified_slice_pool.rs

```diff
@@ -377,7 +377,18 @@ impl Payload {
         message_limit: Option<usize>,
         byte_limit: Option<usize>,
     ) -> CertifiedSliceResult<(Option<Self>, Option<Self>)> {
-        let message_limit = message_limit.unwrap_or(usize::MAX);
+        // `messages_end` may not exceed a certain `max_message_index` in order to cap the number
+        // of signals in the reverse stream. Calculate the maximum number of messages we can include
+        // in the slice such that this is respected.
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_message_index = max_message_index(self.header.begin()).get() as usize;
+            max_message_index.saturating_sub(messages_begin)
+        };
+        let message_limit = message_limit.map_or(max_message_limit, |message_limit| {
+            message_limit.min(max_message_limit)
+        });
```

Comment:
This seems more complicated than necessary, how about something like:
```suggestion
        let max_message_limit = {
            let nr_existing_signals =  self.messages_begin().map_or(0, |mr| mr - self.header.begin()) as usize;
            MAX_STREAM_MESSAGES - nr_existing_signals 
        };
        let message_limit = message_limit.unwrap_or(max_message_limit).min(max_message_limit);
```
(probably doesn't compile, but you get the gist; the last line is arguably not much simpler than what you had before, feel free to ignore that part)